hydra_pywr/nodes/hydropower.py

```diff
@@ -1,37 +1,84 @@
-    AggregatedParameter, ScenarioWrapperParameter
-from pywr.parameters._hydropower import HydropowerTargetParameter
-from pywr.recorders import HydropowerRecorder, NumpyArrayLevelRecorder, NumpyArrayNodeRecorder
-from pywr.domains.river import Catchment
-from pywr.parameters import load_parameter
-import numpy as np
 import pandas as pd
+import numpy as np
+
+from pywr.nodes import (
+    NodeMeta,
+    Link,
+    Input,
+    AggregatedNode,
+    Storage,
+    Output,
+    Catchment
+)
+
+from pywr.parameters import (
+    Parameter,
+    load_parameter,
+    ScenarioWrapperParameter,
+    InterpolatedVolumeParameter,
+    ConstantParameter,
+    HydropowerTargetParameter,
+	AggregatedParameter,
+	MonthlyProfileParameter
+)
+
+from pywr.recorders import (
+    HydropowerRecorder,
+    NumpyArrayNodeRecorder
+)
+
+from pywr.parameters.control_curves import (
+    ControlCurveInterpolatedParameter
+)
+
+# hydra-pywr parameters
 from ..parameters import MonthlyArrayIndexedParameter
 
 import logging
 log = logging.getLogger(__name__)
 
-class LinearStorageReleaseControl(Link):
-    """A specialised node that provides a default max_flow based on a release rule.
-    """
+""" Define nodes exported to `from hydra_pywr.nodes import *` """
+__all__ = (
+    "ProportionalInput",
+    "LinearStorageReleaseControl",
+    "Reservoir",
+    "Turbine",
+    "MonthlyOutput",
+    "MonthlyCatchment"
+)
+
+
+class ProportionalInput(Input, metaclass=NodeMeta):
+    min_proportion = 1e-6
+
+    def __init__(self, model, name, node, proportion, **kwargs):
+        super().__init__(model, name, **kwargs)
+
+        self.node = model.pre_load_node(node)
+
+        # Create the flow factors for the other node and self
+        if proportion < self.__class__.min_proportion:
+            self.max_flow = 0.0
+        else:
+            factors = [1, proportion]
+            # Create the aggregated node to apply the factors.
+            self.aggregated_node = AggregatedNode(model, f'{name}.aggregated', [self.node, self])
+            self.aggregated_node.factors = factors
+
+
+class LinearStorageReleaseControl(Link, metaclass=NodeMeta):
+    """ A specialised node that provides a default max_flow based on a release rule. """
+
     def __init__(self, model, name, storage_node, release_values, scenario=None, **kwargs):
 
-        if release_values is not None:
-            release_values = pd.DataFrame.from_dict(release_values)
+        release_values = pd.DataFrame.from_dict(release_values)
+        storage_node = model.pre_load_node(storage_node)
+
         if scenario is None:
             # Only one control curve should be defined. Get it explicitly
             control_curves = release_values['volume'].iloc[1:-1].astype(np.float64)
             values = release_values['value'].astype(np.float64)
-            storage_node_list = list(filter(lambda x:x.name==storage_node, model.nodes))
-            if len(storage_node_list) != 0:
```

Comment:
Why has this logic been removed? It looks like a sensible check.

Reply:
The `storage_node` is pre_load'ed now on 75, so at this point `storage_node` is a Node ref and not a str as previously.  Rather than use the first matching node by name, the new version is effectively doing the same as the old `else` clause, passing an already-available Node ref to the `ControlCurveInterpolatedParameter`.

If there wasn't a valid `storage_node` arg to the LSRC node, then 75 would catch this when attempting to pre_load, so at the point where this `if` was, we already know there is one matching valid node and have a reference to it.